### PR TITLE
Corrections to combined Latin and Arabic text inside quotes

### PR DESCRIPTION
**Pull request by matthew-ivre**
Before:
image

After:
image

Also added unit tests.

### CODE REVIEW

File: Assets/RTLTMPro/Tests/EngageOrderingTests.cs [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+    public class EngageOrderingTests
+    {
+        // Code to log the encoded output
+        /*
+            foreach (char c in result)
+            {
+                Debug.Log("\\u" + ((int)c).ToString("X4"));
+            }
+        */
+
```
Comment (daniel-ivre) [inline Assets/RTLTMPro/Tests/EngageOrderingTests.cs]: Remove commented code

File: Assets/RTLTMPro/Tests/EngageOrderingTests.cs [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+        // Tests need to replicate what happens in RTLTextMeshPro.GetFixedText()
+        //  If you change the code in that method, make the same changes to this function
+        protected string simulatedUpdateText(string text, bool Farsi, bool FixTags, bool PreserveNumbers)
```
Comment (daniel-ivre) [inline Assets/RTLTMPro/Tests/EngageOrderingTests.cs]: Function names should follow PascalCase naming convention

File: Assets/RTLTMPro/Tests/EngageOrderingTests.cs [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+            return output.ToString();
+        }
+
+        [Test]
+        public void NumberOrder_MidEnd()
+        {
+            // Arrange
+            const string input = "فيديو 360؟ انتقل إلى غرفة 360";
+            const string encodedExpected = "\uFED3\uFBFF\uFEAA\uFBFE\uFEEE\u0020\u0030\u0036\u0033\u061F\u0020\uFE8D\uFEE7\uFE98\uFED8\uFEDE\u0020\uFE87\uFEDF\uFEF0\u0020\uFECF\uFEAE\uFED3\uFE94\u0020\u0030\u0036\u0033";
```
Comment (daniel-ivre) [inline Assets/RTLTMPro/Tests/EngageOrderingTests.cs]: Can't we specify the expected result as a plain string? Is not useful to have it like this when reading the test.

**Review by ruben-ivre — APPROVED** [review]

File: Assets/RTLTMPro/Tests/EngageOrderingTests.cs [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+            rtlTMP.FixTags = true;
+            rtlTMP.ForceFix = false;
+
+            // Act
+            //string result = rtlTMP.UpdateText();
+
+            // Assert
+
+        }
+        */
```
Comment (daniel-ivre) [inline Assets/RTLTMPro/Tests/EngageOrderingTests.cs]: Remove commented code

File: UPMPackage/Tests/EngageOrderingTests.cs
Comment (daniel-ivre) [inline UPMPackage/Tests/EngageOrderingTests.cs:1]: Is this a duplicate? Why?

**Review by daniel-ivre — CHANGES_REQUESTED** [review]

File: Assets/RTLTMPro/Tests/EngageOrderingTests.cs [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+        {
+            var output = new FastStringBuilder(RTLSupport.DefaultBufferSize);
+            RTLSupport.FixRTL(text, output, Farsi, FixTags, PreserveNumbers);
+            output.Reverse();
+
+            return output.ToString();
+        }
+
+        [Test]
+        public void NumberOrder_MidEnd()
```
Comment (ruben-ivre) [inline Assets/RTLTMPro/Tests/EngageOrderingTests.cs]: You can write all these test a single one against a dataset using TestCaseSource and store are these input/expected tests in a single array of string tuples.

**Review by ruben-ivre — APPROVED** [review]

File: Assets/RTLTMPro/Tests/EngageOrderingTests.cs (hunk 3 above)
Comment (matthew-ivre) [reply Assets/RTLTMPro/Tests/EngageOrderingTests.cs]: I can convert them to strings, but I'm not sure that's as useful for comparing outputs. Different applications render the sequences differently. E.g., This is how one of the test results looks in Unity:
image
And this is how the same string looks in Visual Studio:
image

File: UPMPackage/Tests/EngageOrderingTests.cs
Comment (matthew-ivre) [reply UPMPackage/Tests/EngageOrderingTests.cs:1]: Most files are duplicated between the Assets folder and the UPMPackage folder. There were existing tests in both locations. I'm not sure what's required, I've just followed what was there.

File: Assets/RTLTMPro/Tests/EngageOrderingTests.cs (hunk 3 above)
Comment (matthew-ivre) [reply Assets/RTLTMPro/Tests/EngageOrderingTests.cs]: Sorry, that makes sense when they're both presented in the Test Runner:
image

Comment (matthew-ivre) [reply Assets/RTLTMPro/Tests/EngageOrderingTests.cs]: Resolved in latest commit

File: Assets/RTLTMPro/Tests/EngageOrderingTests.cs (hunk 5 above)
Comment (matthew-ivre) [reply Assets/RTLTMPro/Tests/EngageOrderingTests.cs]: Implemented in latest commit

File: Assets/RTLTMPro/Tests/EngageOrderingTests.cs (hunk 2 above)
Comment (matthew-ivre) [reply Assets/RTLTMPro/Tests/EngageOrderingTests.cs]: Resolved in latest commit

**Review by ruben-ivre — APPROVED** [review]

**Review by BenIVRE — APPROVED** [review]

**Review by juanma-ivre — APPROVED** [review]

File: UPMPackage/Tests/EngageOrderingTests.cs
Comment (daniel-ivre) [reply UPMPackage/Tests/EngageOrderingTests.cs:1]: I don't think we should have duplicated tests. @BenIVRE?

Comment (BenIVRE) [reply UPMPackage/Tests/EngageOrderingTests.cs:1]: That's how the package was set up in the first place. However, that being said, the tests don't show up in the test runner once the package is imported. @ruben-ivre do you know if something is configured wrong and that's why they don't? Or is there just really no point in exporting the Test cases with the package?

Comment (ruben-ivre) [reply UPMPackage/Tests/EngageOrderingTests.cs:1]: Apparently tests from packages are not included by default, unless it's an embedded package (stored in the packages folder directly). I guess it makes sense, packages are third party things so you don't need to test them.

There's an option to force that to happen though: https://nagachiang.github.io/unity-showing-tests-from-packages-in-test-runner/

That being said, this repo is horribly organized, duplicating stuff in assets and UPMPackage. It should be either directly a package, or a regular unity project with the package stored in Packages folder. I know it's a fork, but we might consider either making our own or moving stuff around and removing duplicities.

Comment (daniel-ivre) [reply UPMPackage/Tests/EngageOrderingTests.cs:1]: We can point to a specific folder in the repository as well, to avoid including everything

Comment (matthew-ivre) [reply UPMPackage/Tests/EngageOrderingTests.cs:1]: Hi all. Are there changes for me to make here? E.g., delete the Tests subfolder in the UPMPackage folder. Or is that separate work?

**Pull request merged**